### PR TITLE
TYPE: added "Sort by Visibility" option to Structure View

### PR DESCRIPTION
**Pull request by afetisov**
changelog: added "Sort by Visibility" option to Structure View

This is part of an effort to clarify the public APIs of modules. "Sort by Visibility" is a standard option in IDEA plugins, allowing to get a clear view of the public API both at the level of modules and at the level of nested items with more restricted visibility. The main benefit compared to a filter for private items is that one has a clear view of the public API (since those items are sorted to the top), but at the same time keeps the navigation aid for the private API.

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/structure/RsVisibilitySorter.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                    is RsVisibility.Restricted -> Restricted
+                }
+            }
+        }
+    }
+
+    private fun getOrdering(x: Any?): Order {
+        val psi = (x as? RsStructureViewElement)?.value
+        val visibility = (psi as? RsVisible)?.visibility
+        return if (visibility != null) Order.fromVisibility(visibility) else Order.Unknown
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/structure/RsVisibilitySorter.kt:37]: What do you think about considering `#[macro_export]` macros as public?

```suggestion
        return when (val psi = (x as? RsStructureViewElement)?.value) {
            is RsVisible -> Order.fromVisibility(psi.visibility)
            is RsMacro -> if (psi.hasMacroExport && !psi.queryAttributes.isDocHidden) Order.Public else Order.Private
            else -> Order.Unknown
        }
```

**Review by vlad20012 — APPROVED** [review]
:+1: Thanks!
bors r+

**Comment by vlad20012** [comment]
bors retry

File: src/main/kotlin/org/rust/ide/structure/RsVisibilitySorter.kt (hunk 1 above)
Comment (afetisov) [reply src/main/kotlin/org/rust/ide/structure/RsVisibilitySorter.kt:37]: It may be a good idea, but AFAIK currently macros are treated as items with no visibility in the IDE. If we decide to implement it, we'd need to carefully check that the visibility was taken into account in all places.

There are also a couple of edge cases which are currently treated poorly. E.g. a macro `foo!` with no `#[macro_export]` located in `lib::macros` will be treated as private to `lib::macros`. However, we can add `pub(crate) self::foo;` to `lib::macros`, and then `foo` will have Restricted rather than Private visibility (but we cannot add `pub self::foo;` without `#[macro_export]`).

In other words, we would need to properly handle the visibility of macros, and not restrict to a local hack in the Structure View.

Comment (afetisov) [reply src/main/kotlin/org/rust/ide/structure/RsVisibilitySorter.kt:37]: I would also like to add some sort of visibility to Impl items. I'm not certain yet whether it should be projectwide visibility like for macros, or a local hack. The idea is that, while technically an Impl is visible from anywhere, in practice you can only use it if all involved types and traits are visible for you (i.e. you can see the implementing type, all its generic arguments, the trait with its arguments, and all trait bounds on parameters).

**Comment by vlad20012** [comment]
bors retry

File: src/main/kotlin/org/rust/ide/structure/RsVisibilitySorter.kt (hunk 1 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/ide/structure/RsVisibilitySorter.kt:37]: A macro without `#[macro_export]` feels like `pub(crate)` with weird name resolution rules (it can't be accessed outside of the module without `#[macro_use]` or re-export in that module).
I suppose it's better to distinguish public macros from private/restricted ones than just show them at the end of the list.

**Comment by vlad20012** [comment]
I just realized that the implementation is suboptimal because it involves unnecessary name resolution for restricted visibilities (in order to construct `RsVisibility`) when `RsVisStubKind` is enough to solve the problem. The issue is just that `RsVisStubKind` is currently a low-level implementation detail...
